e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -190,67 +196,155 @@ func (pf *PortForwarder) portForwardToPod(pod string, async bool) error {
 	if err != nil {
 		return err
 	}
-	pf.logger.Infof("took %v to wait for port-forward to pod %s/%s (localhost:%d)", took, pf.ns, pod, localPort)
-	pf.setPort(cmd, localPort, pod, async)
+	pf.logger.Infof("took %v to wait for port-forward to pod %s/%s (localhost:%d) (pid:%d)", took, pf.ns, pod, localPort, cmd.Process.Pid)
+	pf.setPort(cmd, localPort, pod)
 	return nil
 }
 
-func (pf *PortForwarder) portForwardToDeployment(async bool) error {
+func (pf *PortForwarder) portForwardToDeployment() error {
 	select {
 	case <-pf.ctx.Done():
 		// Don't bother, subprocess already terminated/terminating
 		return errors.Wrap(pf.ctx.Err(), "failed to start port forward process")
 	default:
 	}
-	// Kill previous port forward process. This isn't strictly necessary since the
-	// subprocess will be killed with the context, and we get a random port every
-	// time. However, this cleans up subprocesses in the interim.
-	if pf.cmd != nil {
-		pf.logger.Infof("stopping port-forward process for %s/%s", pf.ns, pf.deployment)
-		if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
-			return errors.Wrap(err, "failed to kill port forward process")
-		}
-	}
 	pf.logger.Infof("starting port-forward process for %s/%s", pf.ns, pf.deployment)
 	pod, err := pf.kubeClient.GetDeploymentPod(pf.deployment, pf.ns, pf.retryTimeout)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to get deployment pod for %s/%s", pf.ns, pf.deployment))
 	}
-	err = pf.portForwardToPod(pod.Name, async)
+	err = pf.portForwardToPod(pod.Name)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to start port forward for %s/%s", pf.ns, pf.deployment))
 	}
 	return nil
 }
 
-// start creates a port forward to a deployment's pod and keeps it open,
+// Start creates a port forward to a deployment's pod and keeps it open,
 // even if the pod is recreated.
-// Once ctx is cancelled, the port forward will be closed and will no longer be recreated
-func (pf *PortForwarder) start() error {
-	if err := pf.portForwardToDeployment(false); err != nil {
-		return err
+// Once ctx is cancelled, the port forward will be closed and will no longer be recreated.
+// Returns a channel. An error will be sent to the channel, and the channel will
+// be closed, when the PortForwarder stops.
+func (pf *PortForwarder) Start() chan error {
+	pf.mux.Lock()
+	defer pf.mux.Unlock()
+	// defend against the PortForwarder being started more than once
+	if pf.isStarted {
+		panic("PortForwarder already started")
 	}
+	pf.isStarted = true
```

Comment:
Do we want to wait for ready wait here?

In the k8s informer, for example, there's a pattern of waiting until "synced", which in the informer means the List has completed and returned a ResourceVersion. That indicates it's safe to make changes because even if the Watch hasn't started yet, when it starts it will start at the ResourceVersion, so you won't miss any events.

Here, we could block until ready before returning from Start, and that would guarantee that the port-forwarder has come up and the callback was called at least once. It would help guard against race conditions by obviating the need to call InitGitRepos outside the callback, which is inside the PortForwarder lock.

Another option would be to expose a way to wait for ready, either with a WaitForReady func (maybe with ctx cancel to stop waiting) or a ReadyChan method that returns a readyCh which is closed when the PF is next ready (or already closed if it's already ready). I'm also open to adding a new field called `initCh` and having it closed after the first time ready == true, but for that we probably still need a `readyCh`.

Reply:
> which is inside the PortForwarder lock.

Oops. The callback is NOT in the lock, because it calls LocalPort indirectly. But I think that might be ok, as long as we wait for callback completion, rather than just the first ready=true. I guess that means we need to synchronously re-lock after the callback and close the initCh (with sync.Once maybe, since we can't re-close it). Then the Start can block on <-initCh or return the initCh so the caller can.

Reply:
> Then the Start can block on <-initCh or return the initCh so the caller can.

Returning the initCh from Start probably the better option. That way Start doesn't also need to take another context to stop waiting if something happens. The caller can then decide if they want to timeout or do something else async, without needing that complexity in Start.

I guess that means Start is returning an initCh and an errorCh... that makes it seem like maybe they should be combined into an eventCh with an Event struct that can pass InitEvent and ErrorEvent and close when fully done, after being stopped/cancelled/terminated.

Reply:
Done. It now waits for a ready event from the channel